src/fault_log.rs

```diff
@@ -1,134 +1,74 @@
-    /// `Broadcast` received multiple different `Value`s from the proposer.
-    MultipleValues,
-    /// `Broadcast` received multiple different `Echo`s from the same sender.
-    MultipleEchos,
-    /// `Broadcast` received multiple different `Ready`s from the same sender.
-    MultipleReadys,
-    /// `Broadcast` recevied an Echo message containing an invalid proof.
-    InvalidProof,
-    /// `Broadcast` received shards with valid proofs, that couldn't be decoded.
-    BroadcastDecoding,
-    /// `HoneyBadger` could not deserialize bytes (i.e. a serialized Batch)
-    /// from a given proposer into a vector of transactions.
-    BatchDeserializationFailed,
-    /// `DynamicHoneyBadger` received a key generation message with an invalid signature.
-    InvalidKeyGenMessageSignature,
-    /// `DynamicHoneyBadger` received a key generation message with an invalid era.
-    InvalidKeyGenMessageEra,
-    /// `DynamicHoneyBadger` received a key generation message when there was no key generation in
-    /// progress.
-    UnexpectedKeyGenMessage,
-    /// `DynamicHoneyBadger` received a signed `Ack` when no key generation in progress.
-    UnexpectedKeyGenAck,
-    /// `DynamicHoneyBadger` received a signed `Part` when no key generation in progress.
-    UnexpectedKeyGenPart,
-    /// `DynamicHoneyBadger` received more key generation messages from the peer than expected.
-    TooManyKeyGenMessages,
-    /// `DynamicHoneyBadger` received a message (Accept, Propose, or Change)
-    /// with an invalid signature.
-    IncorrectPayloadSignature,
-    /// `DynamicHoneyBadger`/`SyncKeyGen` received an invalid `Ack` message.
-    SyncKeyGenAck(AckFault),
-    /// `DynamicHoneyBadger`/`SyncKeyGen` received an invalid `Part` message.
-    SyncKeyGenPart(PartFault),
-    /// `DynamicHoneyBadger` received a change vote with an invalid signature.
-    InvalidVoteSignature,
-    /// A validator committed an invalid vote in `DynamicHoneyBadger`.
-    InvalidCommittedVote,
-    /// `DynamicHoneyBadger` received a message with an invalid era.
-    UnexpectedDhbMessageEra,
-    /// `BinaryAgreement` received a duplicate `BVal` message.
-    DuplicateBVal,
-    /// `BinaryAgreement` received a duplicate `Aux` message.
-    DuplicateAux,
-    /// `BinaryAgreement` received multiple `Conf` messages.
-    MultipleConf,
-    /// `BinaryAgreement` received multiple `Term` messages.
-    MultipleTerm,
-    /// `BinaryAgreement` received a message with an epoch too far ahead.
-    AgreementEpoch,
-}
+//! Each algorithm can propogate their faulty node logs upwards to a calling algorithm via
+//! `DistAlgorihm`'s `.handle_input()` and `.handle_message()` trait methods.
+
+pub use failure::Fail;
 
 /// A structure representing the context of a faulty node. This structure
 /// describes which node is faulty (`node_id`) and which faulty behavior
 /// that the node exhibited ('kind').
 #[derive(Debug, PartialEq)]
-pub struct Fault<N> {
+pub struct Fault<N, F: Fail> {
```

Comment:
What is the reasoning for replacing the `FaultKind` enum with a generic field? I don't see any clear benefits. 

This method has several drawbacks including the fact that `Fault`s of different kinds would have to be boxed when stored in a collection, etc., not to mention having an extra type parameter everywhere.

What's wrong with nested enums?

Reply:
The intent was to have each module's `FaultKind` defined within the module itself.

If we do want `FaultKind`'s defined in their modules, AND if we want to keep hbbft submodules fairly independent from one another, then I think we have to keep `FaultKind` generic. With nested enums, we still have to identify what type gets nested, which means that we have to import each nested `FaultKind` into the fault_log module, which would mean that every module that uses fault_log indirectly uses every other module. I'm pretty sure that is worth avoiding, right?

I agree that having another generic within `Step` is awkward, so in practice this pull request may not be worthwhile.

A more minor improvement  could be defining submodule `FaultKind`s within the fault_log module and making `FaultKind` enums which nest those values, or we could keep the current behavior where all faultkinds are flat listed. There's no performance improvement with this pull request, it only tries to clarify ownership.

Reply:
I'd just import all of the fault kinds into `fault_log`, renaming each of them to `{module}Fault`, and restore the original `FaultKind` enum. I realize that there will be circular dependencies but it's typical for error handling and isn't a problem.

It would also be fine to move all of the sub-fault kinds into the `fault_log` module. It really doesn't matter.

Reply:
I'd actually prefer the approach with the generic: It's just exactly parallel to the hierarchy of errors. And I don't think there's a situation where we'd ever have to `Box` them: Depending on which algorithm you use, you know which `FaultKind` you get, just like with errors. It avoids unreachable branches in `match` statements and better decouples the modules: There are no `HoneyBadger` variants in the faults returned from `BinaryAgreement`.
Also, we have a type alias for `Step` in each of the algorithms' modules which fixes the additional generic argument, so it doesn't inconvenience the user.

Reply:
Fair enough.